### ISSUE TITLE
Request invoice PDF url

### ISSUE THREAD

**Issue by juanjgarcia** [redguava/cliniko-api#321]
Hi team, it would be really helpful if you could add a URL to the `Invoice` response which when `GET` would offer up the PDF render of that invoice. 

This has multiple applications, from archival, to automated bill submissions to insurance providers.

If there's a workaround for now, let me know, as given a recent change in the UK insurance environment it would now be very useful.

**Comment by bpinto** [comment]
I think this should be doable! I will confirm early next week and reply here.

**Comment by bpinto** [comment]
Well well, it's proven to be harder than I anticipated. There are some things we need to discuss before we can introduce a PDF end point:

1. Would it work via GET or POST? I'm thinking POST could be nice for supporting custom parameters (e.g. margin, page size).

2. We need to update our API to support extra accept headers `application/pdf` as of right now we only support `application/json`.

**Comment by juanjsebgarcia** [comment]
For the time being I've just engineered a solution which interfaces with the front end. 

At a guess the request flow on your side is:
1) `Invoice Object` render to HTML
2) HTML render to PDF

So there's no pre-generated invoice PDFs sat anywhere, they're lazily created on fetch? Hence why we can push params to the generator func?

At any rate, I think an endpoint to get the exact PDF currently being created by the frontend without any modifications would be a great v1. Though I understand that might then compromise the ease of being able to add modifications later whilst maintaining backwards compat.

Long story short, I've worked around it for now, so take your time and do it your way :D

**Comment by DenisBogatirov** [comment]
Is there any way to use API token to "auth user"? What I whant to do is just open https://trainm3.au1.cliniko.com/patients/001/treatment_notes/111.pdf in new tab but currently will receive 401 error

**Comment by bpinto** [comment]
@DenisBogatirov you can't do that, you are authenticated to the API, not the web app.

**Comment by DenisBogatirov** [comment]
@bpinto Is there any way to get link to the file through API? I'm making mobile app and it's crucial to have ability to download files.

**Comment by bpinto** [comment]
@DenisBogatirov you could link to the URL cliniko uses and open the link in a browser, the user will download it if they are logged into the app. If not, they will be asked to login first.

**Comment by DenisBogatirov** [comment]
> @DenisBogatirov you could link to the URL cliniko uses and open the link in a browser, the user will download it if they are logged into the cliniko app. If not, they will be asked to login first.

It will not prompt login, and this app are for patients, so they don't have cliniko credentials

**Comment by juanjsebgarcia** [comment]
You will need to front the request for the patient by either downloading the file, serving it and deleting it, or it might be possible to use a HTTP streaming response.

However not sure if either is permitted or advisable (security/GDPR). 😅

**Comment by Ascyron** [comment]
This would also be really useful for other PDF documents such as patient forms!

Our practice has a legal requirement to download certain forms into another piece of software, and ability to do this via API would save incredible amounts of time!
